### ISSUE TITLE
Unifier dots the relevant pattern variables when it should dot the irrelevant ones

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#2640]
@UlfNorell: I think your abuse of my `..`-modality is buggy:
```agda
-- Andreas, 2017-07-11
-- Bug in compiler in connection with .. and parameter-refinement.

open import Common.IO
open import Common.Bool
open import Common.String

data D : (b : Bool) → Set where
  c : (b : Bool) → D b

test : ..(b : Bool) → D b → Bool
test a (c b) = b

-- With .. the GHC-compiled code prints: Goodbye, world!
-- Without, it prints: Hello, world!

main = putStrLn
  (if (test true (c _)) then "Hello, world!" else "Goodbye, world!")
```

**Comment by UlfNorell** [comment]
Here's an example that's more likely to go wrong (I accidentally get the right behaviour in the original test case):

```agda
open import Agda.Builtin.Nat
open import Common.IO
open import Common.String

data D : (n : Nat) → Set where
  c : (n : Nat) → D n

test : ..(n : Nat) → D n → Nat
test a (c b) = b

main = printNat (test 142 (c _))
```
The problem is that the unifier (cc @jespercockx ) decides to dot the `b` instead of the `a`. If you try to do this yourself you get the expected error:

```agda
test a (c ._) = a -- error: a cannot be used here
```
I'm pretty sure the problem exists not only for my (perfectly reasonable) use of `..`, but also with your abuse of my hijacked notation:

```agda
case_of_ : ∀ {a b} {A : Set a} {B : Set b} → A → (A → B) → B
case x of f = f x

-- Not so shape-irrelevant type!
T : ..(n : Nat) → Set
T n = case test n (c _) of λ where
        zero    → Nat
        (suc _) → String
```
I don't know enough about shape irrelevance to use `T` to break anything, but it doesn't seem like that definition should be accepted.

**Comment by UlfNorell** [comment]
Indeed it should not:

```agda
{-# OPTIONS --experimental-irrelevance #-}
open import Agda.Builtin.Bool
open import Agda.Builtin.Equality

data D : (b : Bool) → Set where
  c : (b : Bool) → D b

test : ..(b : Bool) → D b → Bool
test a (c b) = b

data ⊥ : Set where
record ⊤ : Set where constructor tt

case_of_ : ∀ {a b} {A : Set a} {B : Set b} → A → (A → B) → B
case x of f = f x

-- Not so shape-irrelevant type!
T : ..(n : Bool) → Set
T n = case test n (c _) of λ where
        true  → ⊤
        false → ⊥

data Foo : Set where
  foo : .(b : Bool) → T b → Foo

bad : (b : ⊥) → foo true tt ≡ foo false b → ⊥
bad b _ = b

jackpot : ⊥
jackpot = bad _ refl
```

**Comment by andreasabel** [comment]
This problem exists also for plain irrelevant arguments:
```agda
{-# OPTIONS --experimental-irrelevance #-}

open import Agda.Builtin.Nat
open import Common.IO
open import Common.String

data D : ..(n : Nat) → Set where
  c : (n : Nat) → D n

test : .(n : Nat) → D n → Nat
test a (c b) = b

main = printNat (test 142 (c _))
```
Running the compiled program gives:
```
Issue2640b: internal error: PAP object entered!
    (GHC version 7.10.3 for x86_64_unknown_linux)
    Please report this as a GHC bug:  http://www.haskell.org/ghc/reportabug
Aborted (core dumped)
```
GHC is humble enough to blame it on itself, but looking at the generated code, it is really Agda's fault:
```haskell
name12 = "Issue2640b.test"
d12 v0 v1
  = case coe v1 of
      C8 v2 -> coe v0
      _ -> coe MAlonzo.RTE.mazUnreachableError
main = d18
name18 = "Issue2640b.main"
d18
  = coe
      MAlonzo.Code.Common.IO.d32
      (coe d12 erased (coe C8 (142 :: Integer)))
```
The compiled `test` (`d12`) returns its first argument which has been erased due to irrelevance.

@jespercockx: When you have time again, please modify the unifier to take into account relevance information when distributing the dots.

**Comment by jespercockx** [comment]
This should probably be fixed before the release of 2.5.3. I'll take a look at it now.

**Comment by jespercockx** [comment]
The more general problem here is the following. Since shape-irrelevant terms can be used relevantly at the type level, the pattern matching unifier may encounter equations between shape-irrelevant and relevant terms. Hence if the relevant side is a variable, then it will be solved with a shape-irrelevant solution (which may or may not be a variable). So we need to check that the given term can be used in a relevant position before solving the equation.

**Comment by andreasabel** [comment]
Yeah, this is the general crux with shape-irrelevance.  See also #2595.

**Comment by jespercockx** [comment]
I've implemented the check I outlined above (not pushed yet). This fixes the errors in the compilation as Agda now puts the dot on `a` instead of `b`. However, Ulf's proof of false still goes through. It seems there is something more we need to check.

**Comment by jespercockx** [comment]
After failing to fix this last time, I tried again and finally managed to find the solution with some help by @anuyts. Unfortunately, the fix breaks some examples of runtime-irrelevance in the compiler tests, so I had to turn some indices from relevant to shape-irrelevant in VecReverseIrr.agda and IdentitySmashing.agda. I think this is a consequence of the unfortunate conflation of shape-irrelevance, parametricity, and runtime-irrelevance in the Agda codebase. I'll open a new issue to discuss that broader problem.
